src/lib/es2015.core.d.ts

```diff
@@ -43,7 +43,16 @@ interface Array<T> {
      */
     copyWithin(target: number, start: number, end?: number): this;
 
-    toLocaleString(locales: string | string[], options?: Intl.NumberFormatOptions & Intl.DateTimeFormatOptions): string;
+    /**
+     * Returns a string representation of an array.
+     * Each element is converted to a string using its `toLocaleString` method, then concatenated using the list separator of the current locale.
+     * @param locales Passed as the `locales` parameter to each array element's `toLocaleString` method.
+     * @param options Passed as the `options` parameter to each array element's `toLocaleString` method.
+     */
+    toLocaleString(
+        locales?: string | readonly string[],
+        options?: (T extends { toLocaleString(locales?: any, options?: infer O): string; } ? (options: O) => void : never) extends (options: infer P) => void ? P : unknown,

```

Comment:
See notes towards the bottom of the PR description.

This appears to work well, including on all of the examples in the test suite (see baseline diffs under 31d8a82). The only edge case I can see is that it doesn't resolve nested arrays.

The simpler alternative is to leave the type as `Intl.NumberFormatOptions & Intl.DateTimeFormatOptions`, which covers the builtin `toLocaleString` methods just fine. I'm ambivalent towards the change, but thought it would be worth proposing.